File: basics/02-linkedlist/02-21.cpp
/*
 * leetcode 21
 * merge two ordered linked lists
 *
 */

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode* dummy = new ListNode(-1);
        ListNode* p = dummy;
        ListNode* p1 = l1;
        ListNode* p2 = l2;

        while(p1 != nullptr && p2 != nullptr) {
            if (p1->val > p2->val) {
                p->next = p2;
                p2 = p2->next;
            } else {
                p->next = p1;
                p1 = p1->next;
            }
            p = p->next;
        }
        // in case there are still remaining nodes
	if (p1 != nullptr) p->next = p1;
        if (p2 != nullptr) p->next = p2;

        return dummy->next; 
    }
};

File: basics/02-linkedlist/02-21.py
"""
leetcode 21

this is a recursive solution in python3
"""

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        if not list1:
            return list2
        if not list2:
            return list1
        
        tmp = ListNode()
        if list1.val > list2.val:
            tmp = list2
            tmp.next = self.mergeTwoLists(list1, list2.next)
        else:
            tmp = list1
            tmp.next = self.mergeTwoLists(list1.next, list2)
        return tmp
